### DISCUSSION TITLE
Give players more recognition for passing a map/Implement a "Stage cleared" system

### DISCUSSION THREAD

**Discussion by Mechanizen** [ppy/osu#21721] [Ideas]
Getting a good score usually means getting a good accuracy/good combo/FC/PP. But there is another instance where a score can be qualified as a "good score" which is simply passing a map. 

A good amount of players (including myself) play to try to pass hard maps. We know that we cannot get a "good score" in the common sense (not skilled enough as a player or the map is very hard/playing with a lot of mods). So that for us, simply passing can be qualified as a "good score". That is also similar to the way a new player improves; at first a new player will play 2* maps then 3/4* then attempt to pass their first 5* map, etc...

So we know that simply passing a map is an objective for two kinds of players : new players trying to improve their hardest map passed and experienced players trying to pass maps with very specific skillsets/abnormally hard maps.

The problem is at the moment simply passing a map doesn't really get you any reward. **Obviously I'm not talking about a reward in terms of PP but just a little something that would allow the player to keep track of whether they passed a map or not (also taking mods into account).**

_Before exposing my proposition I have to explain my view on the process of "trying to pass a hard map"._

So say a player wants to try to pass a hard map, first what does the term "hard map" means? When a player qualifies a map as "hard" it's because they know the map is challenging for their skill and that they will struggle to pass it. Thus the player knows in advance that it is very likely that they will fail a lot of times before passing the map and so they have to grind for it. Though simply playing the map and failing over and over on the first difficulty spike is not satisfying so most of the time the player will resort to play with NoFail to be able to play the whole map and improve until they can finally pass it.

But there comes the catch : to improve on a map, the player has to use NoFail but at the same time if the player really passes the map they cannot really say they did because they had NoFail on. But the player has no way of knowing whether they will manage to pass the map on their new try. So the player has to gamble by disabling NoFail and basically say "I will pass the map on this next play".

So the problem is that when a player tries to pass a hard map, the process of grinding to improve and trying to finally get a first pass are the same. You play the map until you finally make it but you have no way to know whether you will make it on your very next play, 

It also causes another problem which is that at the moment the only way for a player to certify that they have passed a map is to submit a play without NoFail. So I think that right now "passing a hard map" is not really well handled by the game.

What I would love to see as a player would be a system that would reward me if I passed a map even if I had NoFail on. So that I don't have to gamble anymore whether I will pass on my next play or not, I could just keep NoFail on play until I finally make it.

That would be possible because what does "failing" mean? "Failing" means that your HP=0, NoFail only prevents the play to end if the player fails but it doesn't prevent the HPs to fall to 0. So it would be great if the game could actually track the HPs throughout the play and reward the player if their HPs never fell to 0. Hence that would rule out the NoFail On/Off gambling and make the grinding process much smoother.

This could take the form of a "Stage Cleared" system which we actually see in a lot other games. Passing the map would mark the play as "cleared" and give it a distinctive check mark with a special animation upon finishing the play. I tried making the following visuals to try to understand how it would turn out (please don't judge my Gimp skills, only the idea lol).

![Song Selection](https://user-images.githubusercontent.com/116296019/208490367-7aab767a-cb66-4a71-81d6-25bb51254643.png)

![Score Recap](https://user-images.githubusercontent.com/116296019/208490759-f4d3ea4d-7708-4bf7-8cbd-184720698da9.png)

On the score recap screen, the "Stage Cleared" thingy could look way better with some additional animations/sound effect to give it more fluff. It looks undrewhelming but you get the idea.

**So to sum it up :**

Current state:
- Both new and experimented players are susceptible to play only to pass maps
- The use of NoFail currently forces the player to repetitively turn it ON and OFF to gamble whether they will pass the map on their next play
- Not dieing on a play with NoFail on "doesn't count"/can't prove the play was a pass unless someone watches the full replay
- The act of passing a map is not rewarded

Proposed changes:
- Make it so that the game detects whether the player's HPs never hit 0 and mark the difficulty as CLEARED (even with NoFail ON)
- Mark the play as CLEARED
- Display a Stage Cleared animation upon finishing the map

Targetted improvements:
- Reward a different way of playing the game
- Give more recognition to the player for improving and grinding for specific scores
- Smooth out the grinding process to pass a map (no longer need to switch NoFail ON and OFF to certify the play was a pass)
- The player is no longer scared to make THE play with NoFail ON
- Make the score recap screen less underwhelming with a cool "Stage Cleared" animation
- Makes it easier for the player to keep track of the maps they already cleared

I haven't talked about whether this is something that should be displayed on the website. I think it could but at first passing a map is a personal achievement so if it could be only a local feature it would be fine.

Also for clarification, even if the player gets a pass with NoFail the 0.5x multiplier stays on, rating a play as a pass does not affect the score.

Let me know what you think!

**Comment by Flamiii** [comment]
I'm not really sure about all of this, but for pass players and those who use NoFail, I think having a good HP graph on the results screen similar to the one currently shown in stable would be nice. Having an enlarged HP graph that explicitly uses some sort of red dot at the bottom to show the exact point in the map where you failed, turning the graph line red to show the portion of the map where your play was considered to be "failed" (similar to stable multiplayer), and then maybe a green dot at the top to show where your HP hit the max and you revived (again, taken from multiplayer). I think something along these lines would be great for players who play for passes.

**Reply by Mechanizen** [reply]
That could be a solution too, it would work fine to show whether a play is failed or not but it doesn't really have a "reward" feeling though

**Comment by peppy** [comment]
It seems the only thing that is wanted here could be summed up in one sentence: Track whether NoFail plays actually failed or not.

Everything else is already tracked.

Not sure about this direction, seems a bit weird. I'd consider not using nofail in the first place.

**Comment by WitherFlower** [comment]
I think clearing a map for the first time is something the level system should reward, by giving some bonus xp for example.
This idea was proposed [here](https://github.com/ppy/osu/discussions/17124) as "total submitted scores".

**Comment by cdwcgt** [comment]
Show gray Nofail in score panel to indicate that NF is not useful(no failed in play) feel feasible?
